feat: Getregion function (#13998)
vim/vim: runtime/doc/builtin.txt

getreg([{regname} [, 1 [, {list}]]])
				String or List   contents of a register
getreginfo([{regname}])		Dict	information about a register

> getregion({pos1}, {pos2})	String	get the text from {pos1} to {pos2}

Reviewer: This function returns a List of strings.  So change "String" -> "List".

		Can also be used as a |method|: >
			GetRegname()->getreginfo()

getregion({pos1}, {pos2}				*getregion()*
		The result is a list of strings where each item is a line

> 		from {pos1} to {pos2}.

Reviewer 2: The docs doesn't mention that `pos1` and `pos2` are swapped if `pos1` is after `pos2`.

Author: OK.  I have added it.

		Can also be used as a |method|: >
			GetRegname()->getreginfo()

> getregion({pos1}, {pos2}				*getregion()*

Reviewer 3: missing closing parenthesis

Author: It is already fixed.

		The result is a list of strings where each item is a line

> 		from {pos1} to {pos2}.  {pos1} and {pos2} are swapped if

Reviewer 3: > The result is a list of strings where each item is a line from {pos1} to {pos2}

I don't understand that. What does this mean.

> {pos1} and {pos2} are swapped if {pos1} is after {pos2}.

Same here, I think you mean, it will always return from the upper left position to the lower right position, right?

Author: > Same here, I think you mean, it will always return from the upper left position to the lower right position, right?

Yes.

		Can also be used as a |method|: >
			GetRegname()->getreginfo()

> getregion({pos1}, {pos2}				*getregion()*

Reviewer 4: Seems a typo.  Closing parenthesis is missing after `{pos2}`.

Author: Fixed.

		The result is a list of strings where each item is a line

> 		from {pos1} to {pos2}.  {pos1} and {pos2} are swapped if

Reviewer 5: Returns the list of strings from {pos1} to {pos2}.

For possible values of {pos1} and {pos2} see |line()|.

Useful to get text starting and ending in different columns, such as |characterwise-visual| selection.

Note that 
    - if |virtualmode| is enabled and selection is past the end of line, result lines are filled with spaces,
    - if selection starts or ends in the middle of multiwidth character (e.g. <tab>), it is not included but it's selected part is substituted with spaces.

Example of getting text of visual selection:
....
Example of getting text of visual-block selection:
....
Example of getting text from arbitrary pos1 and pos2
...
Example of getting text from mark1 and mark2
...

Author: I have updated.

		Can also be used as a |method|: >
			GetRegname()->getreginfo()

getregion({pos1}, {pos2})				*getregion()*
		Returns the list of strings from {pos1} to {pos2}.

> 		For possible values of {pos1} and {pos2} see |line()|.

Reviewer: This function uses the line and column number from the specified position.  The line() function returns only the line number.  I think that should be mentioned here.   The supplied position argument specifies the starting and ending line number and column number.

Author: Updated.

		It is useful to get text starting and ending in different
		columns, such as |characterwise-visual| selection.

		Note that

> 		- {pos1} and {pos2} are swapped if {pos1} is after {pos2}.

Reviewer 5: thx! minor thing about typography (I don't know what vim help prefers), but either way list items should be separated by `,` or `;` and/or `.` for the last list item.

Author: I have checked the other documentation.  It seems terminated by ".".

Reviewer 3: I have mentioned it before, I think this is confusing and I still don't know what is meant with swapping. I think it is more clear to say:

```
Order of {pos1} and {pos2} doesn't matter, it will always return content from the upper left position to the lower right position.
```

		- If |virtualmode| is enabled and selection is past the end of

> 		  line, result lines are filled with spaces.

Reviewer: Suggestion: "resulting lines are filled"

> 		- If selection starts or ends in the middle of multiwidth

Reviewer: Typo: "in the middle of a multibyte"

		Can also be used as a |method|: >
			GetRegname()->getreginfo()

getregion({pos1}, {pos2})				*getregion()*
		Returns the list of strings from {pos1} to {pos2}.
		For possible values of {pos1} and {pos2} see |line()|.
		This function uses the line and column number from the
		specified position.
		It is useful to get text starting and ending in different
		columns, such as |characterwise-visual| selection.

		Note that:
		- {pos1} and {pos2} are swapped if {pos1} is after {pos2}.

> 		- If |virtualmode| is enabled and selection is past the end of

Reviewer: I think you meant "virtualedit" here instead of "virtualmode".

Author: Thanks.  I will fix it.

		  line, result lines are filled.
		- If selection starts or ends in the middle of multibyte
		  character, it is not included but it's selected part
		  is substituted with spaces.

		Examples: >
			:xnoremap <CR>  <Cmd>echo getregion('v', '.')<CR>

> <

Reviewer: Can you also add the following text?

```
    Can also be used as a method:
                  StartPos->getregion(endpos)
```

		Can also be used as a |method|: >
			GetRegname()->getreginfo()

getregion({pos1}, {pos2}, {type})			*getregion()*
		Returns the list of strings from {pos1} to {pos2}.
		For possible values of {pos1} and {pos2} see |line()|.

> 		The {type} is region type:

Reviewer: The third argument should be an optional Dict.  This will allow us to add additional optional arguments as Dict items in the future.  Otherwise, you need to specify all the mandatory arguments.

Author: I don't think `type` should be optional.

The optional arguments should be added later.

Reviewer 3: Agree, the selection type should always be given, no?

> 			"v" for |charwise| mode

Reviewer 2: ```suggestion
			"v" for |characterwise| mode
```

			"V" for |linewise| mode

> 			"<CTRL-V>" for |blockwise-visual| mode

Reviewer: It is not clear to me how the specified visual mode type is used.  I can see that irrespective of the current visual mode and the mode type argument, getregion() always returns the visually selected region of text.  I expected that if a region of text is selected using character-wise visual mode and getregion() is called with "V", then it will return an empty List.

Author: `{pos1}`, `{pos2}` is just used as position.
position `v` means `use visual mode start position`.

Reviewer: I am referring to the "v" or "V" value in the new {type} argument.  For example, you select some characters using character-wise visual mode and then call getregion() with the "V" or "<CTR-V>" for the {type} argument, what value will it return?

Author: It depends on your cursor position.

Reviewer 4: In my previous suggestion (https://github.com/vim/vim/pull/13998#issuecomment-1941481650), the {type} argument is for specifying which style is used to get text from {pos1} to {pos2}.  In other words, if you pass "V" for the {type} argument getregion() will return text from {pos1} to {pos2} as if it's selected in line-wise visual mode regardless of the current visual mode.  (So `getregion({pos1}, {pos2}, "V")` is also equal to `getline(getpos({pos1})[1], getpos({pos2})[1])`.)

At least I tried, the current implementation of getregion() follows this, so I think getregion() call with "V" for the {type} argument should return text of lines between the line of {pos1} and the line of {pos2}.

About getregion() call with "" for the {type} argument, I don't have an absolute one answer but I think the behavior should be one of these:
- Give an error like "{type} argument cannot be empty" or "Invalid value for the {type} argument"
- Use the current visual mode in visual mode and chararacter-wise mode in the other mode
- Use the current visual mode in visual mode and the last visual mode in the other mode

Reviewer: > In my previous suggestion ([#13998 (comment)](https://github.com/vim/vim/pull/13998#issuecomment-1941481650)), the {type} argument is for specifying which style is used to get text from {pos1} to {pos2}. In other words, if you pass "V" for the {type} argument getregion() will return text from {pos1} to {pos2} as if it's selected in line-wise visual mode regardless of the current visual mode. (So `getregion({pos1}, {pos2}, "V")` is also equal to `getline(getpos({pos1})[1], getpos({pos2})[1])`.)
> 
> At least I tried, the current implementation of getregion() follows this, so I think getregion() call with "V" for the {type} argument should return text of lines between the line of {pos1} and the line of {pos2}.
> 

A few tests should be added for this.  Make a characterwise selection, call getregion() with the "v", "V" and "<C-V>" types and check the results.  Do the same for the linewise and blockwise selections.

Reviewer: The help text also needs to be updated to clarify the use of the "type" argument.  Irrespective of the current visual mode, the supplied "type" will be used to determine the region of text returned.

Author: > A few tests should be added for this. Make a characterwise selection, call getregion() with the "v", "V" and "" types and check the results. Do the same for the linewise and blockwise selections.

I have added this.

Author: > The help text also needs to be updated to clarify the use of the "type" argument. Irrespective of the current visual mode, the supplied "type" will be used to determine the region of text returned.

I have updated the documentation.  But the suggestion is wellcome.

		You can get current selection type by |visualmode()|.
		This function uses the line and column number from the
		specified position.
		It is useful to get text starting and ending in different
		columns, such as |characterwise-visual| selection.

		Note that:
		- Order of {pos1} and {pos2} doesn't matter, it will always
		  return content from the upper left position to the lower
		  right position.
		- If 'virtualedit' is enabled and selection is past the end of
		  line, result lines are filled.

> 		- If selection starts or ends in the middle of multibyte
> 		  character, it is not included but it's selected part
> 		  is substituted with spaces.

Reviewer 2: ```suggestion
		- If selection starts or ends in the middle of a multibyte
		  character, it is not included but its selected part is
		  substituted with spaces.
```

		Examples: >
			:xnoremap <CR>

> 			\ <Cmd>echo getregion('v', '.', visualmode())<CR>

Reviewer 2: `visualmode()` returns the type of the last Visual mode, while a `<Cmd>` mapping remains in the current Visual mode, which should be obtained using `mode()`, so this mapping should either using `mode()` instead of `visualmode()` or use `:` instead of `<Cmd>`.

<
		Can also be used as a |method|: >

> 			StartPos->getregion(endpos)

Reviewer 2: Variables in method examples seem to be usually in lower-case (as opposed to user functions), and using lower-case `startpos` here is more consistent with `endpos`.
```suggestion
			startpos->getregion(endpos)
```

		Can also be used as a |method|: >
			GetRegname()->getreginfo()

getregion({pos1}, {pos2}, {type})			*getregion()*
		Returns the list of strings from {pos1} to {pos2} as if it's
		selected in visual mode of {type}.
		For possible values of {pos1} and {pos2} see |line()|.
		The {type} is region type:
			"v" for |characterwise| mode
			"V" for |linewise| mode
			"<CTRL-V>" for |blockwise-visual| mode
		You can get the last selection type by |visualmode()|.
		If Visual mode is active, use |mode()| to get the Visual mode
		(e.g., in a |:vmap|).
		This function uses the line and column number from the
		specified position.
		It is useful to get text starting and ending in different
		columns, such as |characterwise-visual| selection.

		Note that:
		- Order of {pos1} and {pos2} doesn't matter, it will always
		  return content from the upper left position to the lower
		  right position.
		- If 'virtualedit' is enabled and selection is past the end of
		  line, result lines are filled.
		- If selection starts or ends in the middle of a multibyte
		  character, it is not included but its selected part is
		  substituted with spaces.

> 

Reviewer: Typically the error cases for a builtin function are described at the end of the description.  Can you add the conditions under which this function will return an empty List here?  For example, if an unsupported visual mode type is specified, an empty List is returned.  If "v" is specified as the start or end position and the visual mode is not active, then an empty List is returned.  There are other error cases also.

Author: Fixed.
